src/box/box.cc

```diff
@@ -5670,6 +5687,30 @@ box_cfg_xc(void)
 	if (dd_version_id > version_id(2, 10, 1))
 		txn_limbo_filter_enable(&txn_limbo);
 
+	/*
+	 * If we have no space _gc_consumers, wal_cleanup_delay is set to zero
+	 * and we have registered replicas (except for self), we should set
+	 * wal_cleanup_delay to non-zero value to prevent xlogs from being
+	 * deleted.
+	 * Note that we do this after recovery - space _gc_consumers is
+	 * recovered then if it was in snapshot.
+	 */
+	if (!gc_has_persistent_consumers() &&
+	    box_check_wal_cleanup_delay() == 0 &&
+	    replicaset.registered_count > 1) {
+		int wal_cleanup_delay_default = 4 * 3600;
+		say_warn("Current schema does not support persistent WAL GC "
+			 "state, so wal_cleanup_delay option is automatically "
+			 "set to %d to prevent xlogs needed for replicas from "
+			 "being cleaned up. Please re-configure it after "
+			 "schema upgrade if it is needed.",
+			 wal_cleanup_delay_default);
+		gc_set_wal_cleanup_delay(wal_cleanup_delay_default);
+	} else {
+		if (box_set_wal_cleanup_delay() != 0)
+			diag_raise();
+	}
+
```

Comment:
Do we really need to bother disabling this option internally? AFAIU it'll be ignored anyway if all consumers are persisted. Can't we just leave it enabled but print a warning (and raise an error if the compat is set) on any attempt to reconfigure it?

Reply:
If you are talking about the whole approach - it works in this way now:
```c
static double
box_check_wal_cleanup_delay(void)
{
	...
	/* Non-zero and non-default value is detected - say a warning. */
	if (value > 0 && value != default_value)
		say_warn_once("Option wal_cleanup_delay is deprecated.");

	if (value > 0 && wal_cleanup_delay_is_disabled) {
		diag_set(ClientError, ER_DEPRECATED,
			 "Option wal_cleanup_delay");
		return -1;
	}
	...
}
```

Reply:
I meant something like this:

<details><summary>diff</summary>

```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index 993ae1169671..71bfbe790a00 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -3232,6 +3232,13 @@ box_set_wal_cleanup_delay(void)
 	 */
 	if (box_is_anon())
 		delay = 0;
+	if (delay > 0) {
+		say_warn_once("Configuration option 'wal_cleanup_delay' is deprecated");
+		if (wal_cleanup_delay_is_disabled) {
+			diag_set(ER_DEPRECATED, "Configuration option 'wal_cleanup_delay'");
+			return -1;
+		}
+	}
 	gc_set_wal_cleanup_delay(delay);
 	return 0;
 }
diff --git a/src/box/gc.c b/src/box/gc.c
index 0f598bbb7ee5..3538aaea832d 100644
--- a/src/box/gc.c
+++ b/src/box/gc.c
@@ -125,7 +125,7 @@ gc_init(on_garbage_collection_f on_garbage_collection)
 	/* Don't delete any files until recovery is complete. */
 	gc.min_checkpoint_count = INT_MAX;
 
-	gc.wal_cleanup_delay = TIMEOUT_INFINITY;
+	gc.wal_cleanup_delay = 4 * 3600;
 	gc.delay_ref = 0;
 	gc.is_paused = true;
 	say_info("wal/engine cleanup is paused");
diff --git a/src/box/lua/config/instance_config.lua b/src/box/lua/config/instance_config.lua
index 5d06ba27c3b7..66af9ba0bea1 100644
--- a/src/box/lua/config/instance_config.lua
+++ b/src/box/lua/config/instance_config.lua
@@ -1304,7 +1304,7 @@ return schema.new('instance_config', schema.record({
         cleanup_delay = schema.scalar({
             type = 'number',
             box_cfg = 'wal_cleanup_delay',
-            default = 4 * 3600,
+            default = 0,
         }),
         retention_period = enterprise_edition(schema.scalar({
             type = 'number',
diff --git a/src/box/lua/load_cfg.lua b/src/box/lua/load_cfg.lua
index 254c0fd83a5c..31119ded32ce 100644
--- a/src/box/lua/load_cfg.lua
+++ b/src/box/lua/load_cfg.lua
@@ -156,7 +156,7 @@ local default_cfg = {
     wal_max_size        = 256 * 1024 * 1024,
     wal_dir_rescan_delay= 2,
     wal_queue_max_size  = 16 * 1024 * 1024,
-    wal_cleanup_delay   = 4 * 3600,
+    wal_cleanup_delay   = 0,
     wal_retention_period = ifdef_wal_retention_period(0),
     wal_ext             = ifdef_wal_ext(nil),
     force_recovery      = false,
@@ -704,6 +704,7 @@ local dynamic_cfg_skip_at_load = {
     secure_erasing          = ifdef_security(true),
     password_lifetime_days  = ifdef_security(true),
     wal_retention_period    = ifdef_wal_retention_period(true),
+    wal_cleanup_delay       = true,
 }
 
 -- Options that are not part of dynamic_cfg_modules and applied individually
```

</details>

The code looks much simpler and should work just fine AFAIU: initializing `gc.wal_cleanup_delay` to 4h in `gc_init` should be okay because if persistent GC is available, it'll be effectively ignored anyway.

Reply:
Discussed f2f. Agreed to do the following:
1. Mark `wal_cleanup_delay` as `skip_at_load` in `load_cfg.lua`.
2. Set `wal_cleanup_delay` to `nil` in `load_cfg.lua`.
3. Set `wal_cleanup_delay` to `box.NULL` in the new config registry.
4. Set `gc.wal_cleanup_delay` to 3600 in `gc_init`.
5. Depending on the new compat option, either raise an error or log a warning in `dynamic_cfg.wal_cleanup_delay` (`load_cfg.lua`). It's okay to log a warning any time the config option is set (i.e. no need to use `say_warn_once`).

This way, the warning/error will be raised on any attempt to set `wal_cleanup_delay` - no matter via the new or the legacy config. The default value will remain 3600, however it'll be effectively ignored after schema upgrade because all consumers will become persistent.

Reply:
I'm okay with it as long as it works as I expect but we do need to get a new approval from @Totktonada.